src/pedersen.rs

```diff
@@ -1,208 +1,45 @@
-use bitvec::{array::BitArray, order::Lsb0, slice::BitSlice};
-use ff::{Field, PrimeField};
-
-/// The field primitive used by [PedersenHash]
-#[derive(PrimeField)]
-#[PrimeFieldModulus = "3618502788666131213697322783095070105623107215331596699973092056135872020481"]
-#[PrimeFieldGenerator = "7"]
-#[PrimeFieldReprEndianness = "little"]
-pub struct Fp([u64; 4]);
-
-impl Fp {
-    /// Transforms [Fp] into little endian bit representation.
-    fn into_bits(mut self) -> BitArray<Lsb0, [u64; 4]> {
-        #[cfg(not(target_endian = "little"))]
-        {
-            todo!("untested and probably unimplemented: big-endian targets")
-        }
-
-        #[cfg(target_endian = "little")]
-        {
-            self.mont_reduce(
-                self.0[0usize],
-                self.0[1usize],
-                self.0[2usize],
-                self.0[3usize],
-                0,
-                0,
-                0,
-                0,
-            );
-
-            self.0.into()
-        }
-    }
-}
-
-/// A point on an elliptic curve over [Fp].
-#[derive(Clone, Debug, Eq, PartialEq)]
-struct CurvePoint {
-    x: Fp,
-    y: Fp,
-    infinity: bool,
-}
-
-impl CurvePoint {
-    fn identity() -> CurvePoint {
-        Self {
-            x: Fp::zero(),
-            y: Fp::zero(),
-            infinity: true,
-        }
-    }
-
-    fn double(&self) -> CurvePoint {
-        if self.infinity {
-            return self.clone();
-        }
-
-        // l = (3x^2+a)/2y with a=1 from stark curve
-        let lambda = {
-            let two = Fp::one() + Fp::one();
-            let three = two + Fp::one();
-            let dividend = three * (self.x * self.x) + Fp::one();
-            let divisor_inv = (two * self.y).invert().unwrap();
-            dividend * divisor_inv
-        };
-
-        let result_x = (lambda * lambda) - self.x - self.x;
-        let result_y = lambda * (self.x - result_x) - self.y;
-
-        CurvePoint {
-            x: result_x,
-            y: result_y,
-            infinity: false,
-        }
-    }
+use crate::pedersen::field::{
+    FieldElement, FieldElementRepr, PEDERSEN_P0, PEDERSEN_P1, PEDERSEN_P2, PEDERSEN_P3, PEDERSEN_P4,
+};
 
-    fn add(&self, other: &CurvePoint) -> CurvePoint {
-        if self.infinity {
-            return other.clone();
-        }
-        if other.infinity {
-            return self.clone();
-        }
+use ff::PrimeField;
 
-        // l = (y2-y1)/(x2-x1)
-        let lambda = {
-            let dividend = other.y - self.y;
-            let divisor_inv = (other.x - self.x).invert().unwrap();
-            dividend * divisor_inv
-        };
+mod field;
+/// The main hash code used by Starknet.
+///
+/// Contains 251 bits of data and is generated by the [pedersen_hash] function.
+#[derive(Clone, Copy, Debug, PartialEq)]
```

Comment:
URL is missing or somewhere I can't look it up...?

Reply:
They should just work as is. The backticks are just additional `code` formatting AFAIK?